### DISCUSSION TITLE
LVGL - How to store the images on QSPI FLASH?

### DISCUSSION THREAD

**Discussion by dolence** [zephyrproject-rtos/zephyr#66251] [Q&A]
I have an QSPI attached on my board (STM32H757). How can I store image files on this external flash? Also, I'm going to use MCUBOOT in the future for firmware updates and would be nice to use it to write these images to the external flash too.

**Comment by FRASTM** [comment]
Personnaly, I am using the STM32CubeProgrammer (V2.13.0). After selecting the corresponding **E**xternal **L**oader, I just download the binary file at the address of the external flash (0x90000000). 
Which should be equivalent to the following CLI command (here example for stm32h747 disco kit)
`$HOME/STMicroelectronics/STM32CubeProgrammer/bin/STM32_Programmer_CLI -c port=swd -el $HOME/STMicroelectronics/STM32CubeProgrammer/bin/ExternalLoader/MT25TL01G_STM32H747I-DISCO.stldr -d build/hello_world_xip/zephyr/zephyr.signed.bin 0x90000000 -rst`

**Reply by dolence** [reply]
I already did that when I wasn't using Zephyr and using TouchGFX for graphics. I was looking for a way to do this while updating the firmware image. @aunsbjerg reply below gave me some direction to follow. Here is the plan: set a storage partition on external flash as he did and I should be able to use attributes to set a region for the converted images to C files. I will give it a try and give you guys an update. Thank you!

**Comment by aunsbjerg** [comment]
I am working on a similar setup for a project, using an H7 and an external flash for LVGL static stuff and firmware images. My dts file looks something like this:

```devicetree
&octospi1 {
	status = "okay";
	pinctrl-0 = <&octospim_p1_clk_pf10
		     &octospim_p1_ncs_pg6
		     &octospim_p1_io0_pf8
		     &octospim_p1_io1_pd12
		     &octospim_p1_io2_pe2
		     &octospim_p1_io3_pf6>;
	pinctrl-names = "default";

	is25lp032d: ospi-nor-flash@0 {
		compatible = "st,stm32-ospi-nor";
		reg = <0>;
		ospi-max-frequency = <DT_FREQ_M(50)>;
		spi-bus-width = <OSPI_QUAD_MODE>;
		data-rate = <OSPI_STR_TRANSFER>;
		size = <DT_SIZE_M(32)>;
		writeoc = "PP_1_1_4";
		status = "okay";

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			slot1_partition: partition@0 {
				label = "image-1";
				reg = <0x00000000 DT_SIZE_K(896)>;
			};

			scratch_partition: partition@E0000 {
				label = "image-scratch";
				reg = <0x000E0000 DT_SIZE_K(128)>;
			};

			storage_partition: partition@100000 {
				label = "storage";
				reg = <0x00100000 DT_SIZE_K(2048)>;
			};
		};
	};
};

&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		boot_partition: partition@0 {
			label = "mcuboot";
			reg = <0x00000000 DT_SIZE_K(128)>;
			read-only;
		};

		slot0_partition: partition@20000 {
			label = "image-0";
			reg = <0x00020000 DT_SIZE_K(896)>;
		};
	};
};
```

I've found that I need `CONFIG_BOOT_MAX_IMG_SECTORS=256` when building the bootloader application, you may need to fiddle a bit with that.

I haven't gotten around to moving LVGL data to external flash yet, but it _should_ be relatively easy to use in LVGL (see [here](https://docs.lvgl.io/8.3/widgets/core/img.html#image-source)). This uses a file system though, so I haven't figured out yet how to load images to external flash during flashing as suggested by @FRASTM - perhaps a file system can be pre-generated somehow?
